thesis/03-implementation.tex

```diff
@@ -83,23 +83,491 @@ \subsubsection{Backend}
+
+When a Guide is created, it is scheduled onto a task queue and later assembled into an Animated Guide. The processing task invokes Solver to calculate intermediate steps forming a smooth folding animation. 
+
+\medskip 
+Guide processing may terminate with 3 different outcomes.
+
+\begin{itemize}
+	\item An Instruction may be impossible to fold programmatically or realistically. It might occur that the solving algorithm never stops, consuming all available resources. To counteract, a hard timeout has been introduced after which the Guide is marked as timed out - \texttt{TMO}.
+	\item An Instruction file may be malformed or unprocessable, in such scenario the Guide is marked as errored - \texttt{ERR}.
+	\item Finally, a properly folded Instruction is marked as done - \texttt{DON}, and the animation file is persisted in the file system.
+\end{itemize}
+
+
+\begin{figure}[H]
+  \caption{Guide states}
+  \centering
+    \includegraphics[width=0.8\textwidth]{assets/3-guide-states.png}
+\end{figure}
+
+
+
+\medskip
+
+\textbf{Accounts}
+
+Accounts sub-application's responsibilities include, but are not limited to:
+\begin{itemize}
+	\item authentication
+	\item account creation
+	\item password resetting 
+	\item password changing
+\end{itemize}
+
+To fulfill its duties, the Accounts sub-application exposes several REST API endpoints.
+The data model is mostly derived from the built-in Django model increasing the development speed considerably.
+\medskip
+
+ 
+\begin{figure}[H]
+  \caption{Accounts - ERD diagram}
+  \centering
+    \includegraphics[width=\textwidth]{assets/3-erd-accounts.png}
+\end{figure}
+
+
+
+
 \subsection{Algorithms}
 % Ciekawsze algorytmy, aspekty, mechanizmy np. logowanie, indeksowane, cachowanie, synchronizacja, backup, jakieś procesy w tle, jakieś progress bary, jakieś analizy, generowanie warstw GISowych, lokalizacja etc. Jest tu często o czym pisać. %
 
-\subsection{Development environment setup}
+\subsection{Quality assurance}
+
+To ensure the quality of the application, the codebase is continuously linted, tested, built, and delivered. Immediate feedback is available for developers, code reviewers, and operators, reducing the possibility of merging a faulty code into the \texttt{master} branch and therefore using a faulty deployment in the production environment.
+
+\medskip
+The pipeline is implemented using \tech{Github Actions}. \tech{Github Actions} comes bundled with \tech{Github}, allowing for seamless integration with the code repository.
+
+\medskip
+Furthermore, no changes can be pushed directly to the \texttt{master} branch. Every commit needs to be reviewed by at least one other developer before being merged. This policy is enforced using \tech{Github} settings.
+
+\medskip
+Each system component has a dedicated pipeline, tailored to specific requirements. 
+
+\medskip
+Continuous delivery and deployment reduce the risk of deploying an application that has not passed all the required checks in a pipeline.
```

Comment:
Do they reduce the risk or negate it altogether?

Reply:
Negate. Changed.